src/handler.rs

```diff
@@ -577,9 +609,239 @@ where
+        if let Some(inflight_addr) = self.inflight_state.get_inflight_mapping_addr() {
+            unsafe {
+                libc::munmap(
+                    inflight_addr.raw_value() as *mut c_void,
+                    self.inflight_state.get_inflight_mmap_size(),
+                )
+            };
+        }
+
+        let mmap_size = inflight.mmap_size;
+        let mmap_offset = inflight.mmap_offset;
+
+        // mmap the file to the memfd region, unsafe as extern mmap is called
+        let mmap_ptr = unsafe {
+            libc::mmap(
+                std::ptr::null_mut::<c_void>(),
+                mmap_size as usize,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                file.as_raw_fd(),
+                mmap_offset as i64,
+            )
+        };
+
+        if mmap_ptr == ret_val as *mut c_void {
+            self.inflight_state.set_inflight_state(None, None, 0);
+            return Err(VhostUserError::InvalidOperation);
+        } else {
+            self.inflight_state.set_inflight_state(
+                Some(file),
+                Some(GuestAddress::new(mmap_ptr as u64)),
+                mmap_size as usize,
+            );
+        }
+
+        self.set_inflight_region_desc_num(inflight.num_queues, inflight.queue_size);
+
+        Ok(())
+    }
 }
 
-impl<S, V, B: Bitmap> Drop for VhostUserHandler<S, V, B> {
+impl<S, V, B> VhostUserHandler<S, V, B>
+where
+    S: VhostUserBackend<V, B>,
+    V: VringT<GM<B>>,
+    B: NewBitmap + Clone,
+{
+    fn get_inflight_queue_size(&mut self, queue_size: u16) -> usize {
+        let queue_region_size;
+        let descr_state_size;
+        let virtio_features = self.get_features().unwrap();
+
+        if virtio_features & (1 << VIRTIO_F_RING_PACKED) == 0 {
+            // Use descriptor and queue states for split virtqueues
+            queue_region_size = mem::size_of::<QueueRegionSplit>();
+            descr_state_size = mem::size_of::<DescStateSplit>();
+        } else {
+            // Use descriptor and queue states for packed virtqueues
+            queue_region_size = mem::size_of::<QueueRegionPacked>();
+            descr_state_size = mem::size_of::<DescStatePacked>();
+        }
+        queue_region_size + descr_state_size * queue_size as usize
+    }
+
+    fn memfd_alloc(&self, mmap_size: usize) -> VhostUserResult<(*mut c_void, i64, i64)> {
+        let mut ret_val;
+        let inflight_file;
+        let dup_inflight_file;
+
+        // create an anonymous file and return a file descriptor to it.
+        // unsafe as extern syscall funciton is called
+        ret_val = unsafe {
+            libc::syscall(
+                libc::SYS_memfd_create,
+                &std::ffi::CString::new("inflight-region").unwrap(),
+                libc::MFD_ALLOW_SEALING,
+            )
+        };
+
+        if ret_val == -1 {
+            return Err(VhostUserError::MemFdCreateError);
+        }
+
+        inflight_file = ret_val;
```

Comment:
inflight_file is a RawFd, so it need to be manually closed on error handling path. Otherwise it will get leaked.

Reply:
The inflight_file is actually a RawFd, and it's error prone to deal with RawFd. 
So it would be better to convert the infight_file from RawFd into a File object, which eases the error handling cases.